forem/forem
# Add co_author_ids to Articles

## What type of PR is this? (check all applicable)


## Description
This PR replaces `second_user_id` and `third_user_id` with a `co_author_ids` column on the `Articles` table to make adding co-authors to posts more flexible in the future. Additionally, this PR adds two methods to the `article_decorater`, `#co_authors`, and `#co_author_name_and_path`, that are used to display the linked co-author names in a human-readable format in the `articles/show.html.erb` view, replaces `second_user_id` and `third_user_id` with `co_author_ids` when necessary in the codebase, refactors existing specs to use `co_author_ids` in place of `second_user_id` and `third_user_id`, adds a `data_update` script to update the `co_author_ids` column with the existing `second_user_id` and `third_user_id` data, and adds validations around co-author IDs not being the same as the author's ID or as a co-author's ID.

## Related Tickets & Documents
Relates to a private issue and relates to PR #10252 . The second part to this PR (the removal of `second_user_id` and `third_user_id` from the Articles table) can be found here: PR #10383 

## QA Instructions, Screenshots, Recordings
**To QA this refactor, ensure that you have `super_admin` permissions and then do the following:**

- Navigate to `/admin/articles` and add `co_author_ids` to the post of your choice (_heads-up_: I created my own post for testing purposes) and click "submit" to save your changes. Upon submit, you should be redirected to the article show page, should see a flash messge indicating a successful update, and you should now see the co-author IDs listed in the view:
<img width="1351" alt="Screen Shot 2020-09-15 at 2 57 29 PM" src="https://user-images.githubusercontent.com/32834804/93264134-ed1ac900-f763-11ea-96b5-523ec540bea8.png">
<img width="1376" alt="Screen Shot 2020-09-15 at 2 57 53 PM" src="https://user-images.githubusercontent.com/32834804/93264202-fefc6c00-f763-11ea-9300-47cb2d0fddd1.png">
<img width="1350" alt="Screen Shot 2020-09-17 at 1 43 36 PM" src="https://user-images.githubusercontent.com/32834804/93520356-cd161180-f8eb-11ea-9510-4473f4c6a4d4.png">

- Click on the post with the newly added co-authors and ensure that they are displayed correctly (in a readable sentence) and that the correct path is associated with each co-author (you should be taken to each individual user's profile upon click):
<img width="907" alt="Screen Shot 2020-09-17 at 1 44 28 PM" src="https://user-images.githubusercontent.com/32834804/93520648-0b133580-f8ec-11ea-842c-e62e3120dec4.png">
<img width="1415" alt="Screen Shot 2020-09-17 at 1 45 42 PM" src="https://user-images.githubusercontent.com/32834804/93520765-3dbd2e00-f8ec-11ea-9933-cfaafb33fc4d.png">

- If you're feeling ambitious, you can add even _more_ co-authors to the post by returning to `/admin/articles` and following the first two QA steps again:
<img width="1351" alt="Screen Shot 2020-09-17 at 1 47 29 PM" src="https://user-images.githubusercontent.com/32834804/93520924-7ceb7f00-f8ec-11ea-8716-dba2c9735c18.png">
<img width="1342" alt="Screen Shot 2020-09-17 at 1 47 54 PM" src="https://user-images.githubusercontent.com/32834804/93520934-807f0600-f8ec-11ea-8b3e-b9fc225acccc.png">
<img width="893" alt="Screen Shot 2020-09-17 at 1 48 03 PM" src="https://user-images.githubusercontent.com/32834804/93520937-8248c980-f8ec-11ea-981c-78f7876a21d2.png">
<img width="1342" alt="Screen Shot 2020-09-17 at 1 48 22 PM" src="https://user-images.githubusercontent.com/32834804/93520940-8379f680-f8ec-11ea-8c3c-d76a0d9f02c8.png">

- To test the co-author-related validations, attempt to add the author of the post as a co-author of the post. If you try to make the author of the post a co-author as well, you should be shown a flash message stating that you cannot do this. Additionally, test that you cannot add the same co-author to a particular article twice and that adding text for the co_author_id is rendered as invalid. You should again, be shown a flash message stating that you cannot do this:
<img width="1342" alt="Screen Shot 2020-09-17 at 1 49 42 PM" src="https://user-images.githubusercontent.com/32834804/93521223-fd11e480-f8ec-11ea-85d8-72f2adc19236.png">
<img width="1349" alt="Screen Shot 2020-09-17 at 1 54 38 PM" src="https://user-images.githubusercontent.com/32834804/93521485-5d088b00-f8ed-11ea-83ab-8c35e5670fce.png">
<img width="1350" alt="Screen Shot 2020-09-17 at 2 03 35 PM" src="https://user-images.githubusercontent.com/32834804/93522410-a73e3c00-f8ee-11ea-80f8-4e97f4736854.png">
<img width="1334" alt="Screen Shot 2020-09-17 at 2 03 45 PM" src="https://user-images.githubusercontent.com/32834804/93522421-aad1c300-f8ee-11ea-9a27-29f3a5e245cc.png">
<img width="1293" alt="Screen Shot 2020-09-30 at 11 14 25 AM" src="https://user-images.githubusercontent.com/32834804/94718091-71e21700-030e-11eb-891e-a4381c5c3e9d.png">
<img width="1312" alt="Screen Shot 2020-09-30 at 12 57 41 PM" src="https://user-images.githubusercontent.com/32834804/94727889-99d87700-031c-11eb-9453-8c47ebabef95.png">

- Finally, try to break things! 🔨 

**_Additionally, please make sure that all of the specs pass and that the app looks as expected._**

## Added tests?

- [x] yes and refactored existing specs to account for changes

## Added to documentation?


## [optional] Are there any post-deployment tasks we need to perform?
Yes: I have added a `data_update` script to update the `co_author_ids` column with the existing `second_user_id` and `third_user_id` data.

## [optional] What gif best describes this PR or how it makes you feel?

![typing](https://media.giphy.com/media/pOZhmE42D1WrCWATLK/giphy.gif)
